### PR TITLE
[FLINK-23946][clients] Application mode should shutdown gracefully on…

### PR DESCRIPTION
**Pull request by dmvk**
https://issues.apache.org/jira/browse/FLINK-23946

### CODE REVIEW

File: flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java
Comment (tillrohrmann) [inline flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java:833]: This test doesn't seem to belong into the `ApplicationDispatcherBootstrapTest`. Maybe it would be better suited for `ApplicationDispatcherBootstrapITCase`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:714]: Why do we want to shut down the cluster exceptionally? Wouldn't it be good enough in such a case to call the fatal exception handler and let it do its magic?

File: flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java:256]: This looks a bit as if we are leaking implementation details for testing purposes. I am not sure whether this really good given that the `MiniCluster` is not a testing utility only. I'd rather go w/o if possible.

File: flink-tests/src/test/java/org/apache/flink/test/runtime/leaderelection/ZooKeeperLeaderElectionITCase.java
Comment (tillrohrmann) [reply flink-tests/src/test/java/org/apache/flink/test/runtime/leaderelection/ZooKeeperLeaderElectionITCase.java:149]: Did the test become unstable because of your previous change? If yes, what is causing the instability? I assume it has something to do that we do the leader election for the `Dispatcher` and the `ResourceManager` individually. Isn't this also a problem for Flink in general if we have two processes where one has the leading `ResourceManager` and the other the leading `Dispatcher`? cc @xintongsong this might have been introduced with FLINK-21667.

It seems that we have the assumption in the `DispatcherResourceManagerComponent` that we need a `ResourceManager` to deregister the application. Before, I think it worked because the `ResourceManager` would also do the deregistering w/o having the leadership. One way to solve this problem properly, is to do a single leader election for the JobManager process.

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrap.java
Comment (tillrohrmann) [inline flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrap.java:172]: Why is it better to shut down the cluster exceptionally than failing hard here?

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrap.java
Comment (tillrohrmann) [inline flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrap.java:173]: I think this does not really solve the described problem. If the `Dispatcher` loses leadership, then it will call `ApplicationDispatcherBootstrap.stop` which will cancel the `applicationCompletionFuture`. In this case, I think we should neither fail nor shut down the cluster because we only want to stop the `ApplicationDispatcherBootstrap` and maybe restart it once we have regained the leadership.

**Review by tillrohrmann — CHANGES_REQUESTED** [review]
Thanks for creating this PR @dmvk. I think we are not fully solving the described problem with this change. Please take a look at my comments.

I also think that we might have uncovered a problem with the `ResourceManagerService` and the deregistration of the applicaiton in the presence of multiple standby JobManager processes.

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrap.java
Comment (tillrohrmann) [reply flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrap.java:172]: I actually think that this might lead to an undesired behaviour. In case of an unexpected failure we would like to restart the process. However, if we shut the cluster down, then it won't get restarted (e.g. on Yarn) because we deregister the application.

File: flink-tests/src/test/java/org/apache/flink/test/runtime/leaderelection/ZooKeeperLeaderElectionITCase.java
Comment (dmvk) [reply flink-tests/src/test/java/org/apache/flink/test/runtime/leaderelection/ZooKeeperLeaderElectionITCase.java:149]: Thanks for creating an issue to track this! This seems to be a bigger issue than I initially thought 🤔 

This instability came from asserting that `DispatcherResourceManagerComponent` shutdownFuture doesn't complete exceptionally.

Screen Shot 2021-08-30 at 8 21 58 AM

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrap.java
Comment (dmvk) [reply flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrap.java:173]: Hmm, I think I may have misinterpreted the issue description then 🤔 I initially thought that don't want any exceptions (even from the user code) to shut down the process fatally.

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrap.java
Comment (dmvk) [reply flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrap.java:172]: You're right. Another approach could be giving up leadership / restarting dispatcher, but this could be tricky to do 🤔 Not sure if that would actually improve anything, probably not worth the effort.

File: flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java
Comment (dmvk) [reply flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java:833]: 👍 Also I'll change the test to reflect described scenario (dispatcher loses leadership).

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationMiniCluster.java
Comment (dmvk) [reply flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationMiniCluster.java:36]: Agreed, but this is hard to do without breaking mini cluster public API 🤔

I did attempt with AbstractMiniCluster and it didn't look right. 😢 Huge change-set that is breaking public API (eg. enum for HAService would have to move to abstract class, configurations, ...). I think this could be done properly, but it would definitely take some time (~ 2MD?), there should be an easier way.

What I can do is to allow passing custom DispatcherResourceManagerComponentFactory into `TestingMiniCluster` via. User would be still able to access methods like `submitJob`, but that would be no different from real world scenario, where this methods are accessible via RPC.

**Comment by dmvk** [comment]
@tillrohrmann Thanks for the review! I've incorporated your suggestions, PTAL

File: flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java
Comment (dmvk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java]: removed

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java
Comment (dmvk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java]: removed

**Review by tillrohrmann — COMMENTED** [review]
Thanks for updating this PR @dmvk. I think the PR is almost ready to be merged. One last question is whether the `MiniCluster` should stop the application if the `shutDownFuture` completes exceptionally. I think this depends on the semantics of this future. If the future is only completed exceptionally iff an unexpected exception occurred, then it might be better to call the fatal exception handler.

File: flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java
Comment (dmvk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java:462]: oh, this is a leftover from the previous approach. I think we currently never complete shutdownFuture exceptionally, so we can safely ignore this and let the `assertNoException` take care of it in case it happens 👍

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrap.java
Comment (dmvk) [reply flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrap.java:175]: :+1:

File: flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapITCase.java
Comment (dmvk) [reply flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapITCase.java:54]: 👍 I always forget this, thinking whether we can cover this by an automated check 🤔

**Comment by dmvk** [comment]
ZooKeeperLeaderElectionITCase is still flaky. I'm not sure if I can stabilize it without actually solving https://issues.apache.org/jira/browse/FLINK-24038

**Comment by dmvk** [comment]
@tillrohrmann I've addressed rest of your concerns. I think we should merge this after https://issues.apache.org/jira/browse/FLINK-24038 is fixed. WDYT?

**Comment by tillrohrmann** [comment]
Sounds good. The problem should not block the release because it was not introduced recently.

File: flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java
Comment (tillrohrmann) [reply flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java:403]: Please double check whether this test method name is correct.

**Review by tillrohrmann — APPROVED** [review]
Thanks for updating this PR @dmvk. The changes look good to me. I had one comment left. +1 for merging after https://issues.apache.org/jira/browse/FLINK-24038 has been fixed.

File: flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java
Comment (dmvk) [reply flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java:403]: Thanks, I've missed this one 👍 Fixed.

**Comment by dmvk** [comment]
@flinkbot run azure

**Comment by dmvk** [comment]
Rebased on top of @nicoweidner's changes + incorporated https://github.com/apache/flink/pull/18136

**Pull request closed without merging**
